### PR TITLE
Feature/issue 558 windows build script

### PR DESCRIPTION
**Pull request by m0rkeulv**
Build script for Windows (Using Ant tasks to download and prepare dependencies)

### CODE REVIEW

File: prepare.xml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    </target>
+
+    <target name="psi.extract">
+        <copy file="${psiFile}" tofile="${basedir}/light-psi-all.jar"/>
+    </target>
+
+    <!--CHECK IF READY TO BUILD-->
+    <target name="ready">
+        <condition property="isReady">
+            <and>
```
Comment (EricBishton) [inline prepare.xml]: This is all that's necessary to build.  If we want to work with the lexers, though, we need the `idea-flex.skeleton` file, which comes from the IDEA community-edition sources: https://github.com/JetBrains/intellij-community/blob/master/tools/lexer/idea-flex.skeleton.  It would be nice to get that into place so that we can eventually remove the generated lexer files and only keep the .flex files in the source tree.

File: prepare.xml [hunk 2]
```diff
@@ -0,0 +1,116 @@
+<project name="Idea Haxe language scripted package plugin" default="init" basedir=".">
+
+
+    <property name="grammarKitVersion" value="1.2.0"/>
+    <property name="toolsDir" value="${basedir}/tools/"/>
+
+    <property name="ideaVersion" value="${version}"/>
+    <property name="ideaIUFile" value="${toolsDir}/ideaIU-${ideaVersion}.tar.gz"/>
```
Comment (EricBishton) [inline prepare.xml]: So, this downloads the linux build.  Should we, instead, download the platform-specific build.  (e.g. .zip for Windows, .dmg for macOS)?

File: fetchGrammarKit.bat [hunk 3]
```diff
@@ -0,0 +1,3 @@
+@Echo Off
+:: call ant tasks that downloads and extracts grammarKit
+call ant -f prepare.xml makeToolsDir grammarKit.fetch grammarKit.extract psi.fetch  psi.extract cleanup
```
Comment (EricBishton) [inline fetchGrammarKit.bat]: I prefer a single ant target to do wrap all of this up.  I think that it will make it easier to maintain on multiple platforms.

File: fetchIdea.bat [hunk 4]
```diff
@@ -0,0 +1,10 @@
+@Echo Off
+if "%1"=="" (
+echo 'This script must be called with the version of IDEA to build'
+echo 'example: build.bat 13.1.6'
+echo 'example: build.bat 2017.2.4'
+Exit /B 1
+)
+
+:: call ant tasks that downloads and extracts IDEA source files
+call ant -f prepare.xml -Dversion="%1" makeToolsDir ideaIU.fetch ideaIU.extract cleanup
```
Comment (EricBishton) [inline fetchIdea.bat]: Again, I prefer a single-target approach.

**Comment by EricBishton** [comment]
Nice work.  Thanks.  Do you have the platforms to test this upon?  And, are you interested in making this ant task replace the shell scripts we have for Linux?

Also, there are a couple of niceties that I've been thinking about and just don't have time to do: Make the scripts into an AppVeyor compatible windows build; Add a "contributor setup" script that adds the proper SDKs (and their libs) to the dev's IDEA installation (basically, automating setup).  Are you interested in working on those?  

(There are other pressing issues that can also use your input, and you are welcome to work where you like.  I'm trying to understand and tap into your interests.  :-)

File: prepare.xml (hunk 2 above)
Comment (m0rkeulv) [reply prepare.xml]: i could get the windows version for windows and unzip but i don't know of any ant task that unpacks .dmg files. i do however think it would be a bit of over engineering, as far as i know we only use the java files which are cross-platform anyway. (i use the Linux files when building on windows without any problem)

Comment (EricBishton) [reply prepare.xml]: Yes, and no.  For the time being, it's OK, because we can generally use the Java that is running on a system.  The default installation of Java is what we're using for Travis, too.

For the long term, we should be using the version of Java that JetBrains delivers as part of their release.  That version specially modified to fix bugs that JetBrains' products run into, and are platform-specific.

File: prepare.xml (hunk 1 above)
Comment (m0rkeulv) [reply prepare.xml]: I could probably add a step that fetches that file and add it to the workspace, I just need to know where to put it and if we are always going to get it from the master branch or if we should use a more "stable" branch where we won’t have to worry about major changes like package/folder refactoring etc.

Comment (EricBishton) [reply prepare.xml]: It goes into the same place that the grammar-kit jar goes: ${basedir}.

The best thing to do is pick it up from the tag matching the build we are making.  It's not terribly easy, though:  After unpacking the idea build, check the file `build.txt`.  It's in the root directory of the package on Linux and Windows.  The content of that file looks like `IU-172.3968.16`.   Dropping the leading portion of that, leaves us with the build number, which, when prepended with `idea/` gives us the git tag name for the build: `idea/172.3968.16`.  However, github doesn't allow us to reach a file by tag (just a source zip).  But, JetBrains always creates a branch for each release.  We can use that branch, which should be close enough, like so: 
`https://github.com/JetBrains/intellij-community/blob/172.3968/tools/lexer/idea-flex.skeleton`.

Comment (m0rkeulv) [reply prepare.xml]: Ok ill give it a shot, but i might need a few days :)

Comment (EricBishton) [reply prepare.xml]: Sure.  No worries.

Comment (EricBishton) [reply prepare.xml]: Just remembered that there is some code to deal with the build version in https://github.com/HaxeFoundation/intellij-haxe/blob/master/common.xml#L47-L148.  I'm not sure if you can re-use it, but it's there for reference, anyway.

**Comment by EricBishton** [comment]
Seems like there is either a problem with the download or the extraction of IDEA, sometimes.

**Comment by EricBishton** [comment]
I've restarted the two failed builds to see if it was a transient error.  If those work, I'll restart a full build and see if the problems recur.

**Comment by m0rkeulv** [comment]
Making an Ant target with this many different outcomes in plain Ant is getting unnecessary complex and painful to maintain due to the fact that properties are immutable, do you have anything against using a third-party lib, more specifically ant-contrib.
It used to be on https://sourceforge.net/projects/ant-contrib/files/ant-contrib/ the page does not load for me; however a lot of people on github has forks. for instance:
https://github.com/cniweb/ant-contrib

**Comment by EricBishton** [comment]
Well, sourceforge is almost dead.  So I don't want to be downloading from there.

I don't mind third-party things, in general.  We can add a tools directory to hold them and include them in the repo.  And, maybe we should just do that and check in the versions of grammar-kit and the skeleton that we need?  Then we wouldn't have to do so much work.

As far as the amount of plain ant code your are using, why don't you use JavaScript to do your complex logic, and once you've figured everything outm just set the properties that the final tasks need?  There is an example of that in common.xml.

**Comment by EricBishton** [comment]
I have no new comments.  Let me know if this is ready to merge or if you want to do more work on it.

**Comment by m0rkeulv** [comment]
I'll give JavaScript a try first, and then maybe ant-contrib. I just don't want it the way it is now as it’s hard to follow. It started out ok as a plain ant job, but along the way the complexity grew and in the end it ended up quite messy.

**Comment by EricBishton** [comment]
@m0rkeulv This has been hanging out for a while.  Do you have any updates, or shall I go ahead and merge it?

**Comment by m0rkeulv** [comment]
@EricBishton  Sorry about the delay, been without a proper computer for a while due to a failing motherboard, ill try to get this stuff done during the weekend :)

**Comment by m0rkeulv** [comment]
Since I am not quite happy with the Ant script I have been looking into alternatives. Gradle seems to have most of the same features and use the same “task” concept as Ant but is more flexible. Jetbrains have also made a plugin that seems to automatically download the necessary IntelliJ source files for you and it probably will also extract the files and do the build job, but I have not gotten that far in my testing yet. Before I invest more time into this I just want to know what you think if porting the possibility of porting build script to Gradle.

https://www.jetbrains.org/intellij/sdk/docs/tutorials/build_system/prerequisites.html

**Comment by EricBishton** [comment]
Personally, I don't mind the thought of using gradle, though I haven't to date.  Make, ant, maven, gradle -- they're all the same to me.  

The requirements are:
 - Works on every platform (Linux, MacOS, and Windows w/o Cygwin).
 - Minimum extra packages (optimally, zero) required to build.
 - Can target specific IDEA versions for debugging.
 - Can build multiple versions of the project.
 - Can set separate build variables/properties per version (like we do with idea_XX.properties).
 - Single command to build the default version (for new users).
 - Automatable for CI (Travis-CI and/or AppVeyor).  Docker builds are OK.
Nice to have:
 - Be able to automate installation of all required packages on a user workstation.
 - Be able to build and run tests for multiple target IDEA versions with a single command.
 - Be able to target versions of IDEA before 2016, though we may not need this.

For deployment, I only care about having different versions for each year, until there is an incompatibility between them.  By publishing separate plugins per year, we can track -- at a gross level -- what versions of IDEA our users are on and how much they are lagging JetBrains' releases.

**Comment by m0rkeulv** [comment]
Due to lack of time I am reverting this back to Ant build scripts, while it’s far from an elegant solution, it works on all OS and can be used until I find the time to continue my Gradle port.
